File: Java/Trees/interval_tree.java
public class interval_tree {
    static class Interval {
        int low, high;

        Interval(int low, int high) {
            this.low = low;
            this.high = high;
        }
    }

    static class Node {
        Interval interval;
        int maxHigh;
        Node left;
        Node right;

        Node(Interval interval) {
            this.interval = interval;
            this.maxHigh = interval.high;
            this.left = null;
            this.right = null;
        }
    }

    static Node insert(Node root, Interval interval) {
        if (root == null)
            return new Node(interval);

        int low = root.interval.low;

        if (interval.low < low)
            root.left = insert(root.left, interval);
        else
            root.right = insert(root.right, interval);

        if (root.maxHigh < interval.high)
            root.maxHigh = interval.high;

        return root;
    }

    static boolean overlap(Interval i1, Interval i2) {
        return i1.low <= i2.high && i2.low <= i1.high;
    }

    static Interval search(Node root, Interval interval) {
        if (root == null)
            return null;

        if (overlap(root.interval, interval))
            return root.interval;

        if (root.left != null && root.left.maxHigh >= interval.low)
            return search(root.left, interval);

        return search(root.right, interval);
    }

    static void inorder(Node root) {
        if (root == null)
            return;

        inorder(root.left);
        System.out.println("[" + root.interval.low + ", " + root.interval.high + "] max = " + root.maxHigh);
        inorder(root.right);
    }

    public static void main(String[] args) {
        Interval[] intervals = { new Interval(15, 20), new Interval(10, 30), new Interval(17, 19),
                new Interval(5, 20), new Interval(12, 15), new Interval(30, 40) };

        Node root = null;
        for (Interval interval : intervals)
            root = insert(root, interval);

        System.out.println("Inorder traversal of constructed Interval Tree is:");
        inorder(root);

        Interval searchInterval = new Interval(6, 7);
        System.out.println("\nSearching for interval [" + searchInterval.low + ", " + searchInterval.high + "]");
        Interval result = search(root, searchInterval);
        if (result != null)
            System.out.println("Overlaps with [" + result.low + ", " + result.high + "]");
    }
}

File: C++/Trees/interval_tree.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Interval
{
    int low, high;
};

struct Node
{
    Interval interval;
    int max_high;
    Node *left;
    Node *right;
};

Node *newNode(Interval i)
{
    Node *node = new Node;
    node->interval = i;
    node->max_high = i.high;
    node->left = nullptr;
    node->right = nullptr;
    return node;
}

Node *insert(Node *root, Interval i)
{
    if (root == nullptr)
        return newNode(i);

    int low = root->interval.low;

    if (i.low < low)
        root->left = insert(root->left, i);
    else
        root->right = insert(root->right, i);

    if (root->max_high < i.high)
        root->max_high = i.high;

    return root;
}

bool overlap(Interval i1, Interval i2)
{
    return (i1.low <= i2.high && i2.low <= i1.high);
}

Interval *search(Node *root, Interval i)
{
    if (root == nullptr)
        return nullptr;

    if (overlap(root->interval, i))
        return &(root->interval);

    if (root->left != nullptr && root->left->max_high >= i.low)
        return search(root->left, i);

    return search(root->right, i);
}

void inorder(Node *root)
{
    if (root == nullptr)
        return;

    inorder(root->left);
    cout << "[" << root->interval.low << ", " << root->interval.high << "] max = " << root->max_high << endl;
    inorder(root->right);
}

int main()
{
    vector<Interval> intervals = {{15, 20}, {10, 30}, {17, 19}, {5, 20}, {12, 15}, {30, 40}};

    Node *root = nullptr;
    for (const auto &interval : intervals)
        root = insert(root, interval);

    cout << "Inorder traversal of constructed Interval Tree is" << endl;
    inorder(root);

    Interval searchInterval = {6, 7};
    cout << endl
         << "Searching for interval [" << searchInterval.low << ", " << searchInterval.high << "]" << endl;
    Interval *result = search(root, searchInterval);
    if (result != nullptr)
        cout << "Overlaps with [" << result->low << ", " << result->high << "]" << endl;

    return 0;
}
